### PR TITLE
Adds progress image to the local source create

### LINKED ISSUE #225: Improve feedback in `--local-path` workflow by showing the size of the local source being uploaded
### Description of problem

In the `--local-path` workflow, there is no way to know what is the size of the local source bundle that is being uploaded.

Screen Shot 2022-07-17 at 11 08 52 PM

### Proposed solution

```gherkin
Given the developer is using the --local-path workflow to create workloads
When they run tanzu apps workload apply/create with --local-path and upload their local source to a registry
Then the size of the local source bundle that is being uploaded, and its progress is shown in the output
```

#### Bonus
- Add a progress bar to show the progress of the upload.

### Additional context
You can use the following imgpkg feature to get size uploaded information.
https://github.com/vmware-tanzu/carvel-imgpkg/blob/develop/pkg/imgpkg/registry/with_progress.go

cc @danfein for the design recommendation on how to display that information.

### PR DESCRIPTION
**Pull request by shaheerkootteeri**
Pull request

### What this PR does / why we need it
PR adds progress bar with image size to the `--local-path `workflow for workload create and apply commands.

### Which issue(s) this PR fixes

Fixes #225 

### Describe testing done for PR

Installed in local and validated against the local cluster

### CODE REVIEW

**Comment by shaheerkootteeri** [comment]
image

**Comment by shaheerkootteeri** [comment]
image

File: pkg/commands/workload.go [hunk 1]
```diff
@@ -429,11 +429,13 @@ func (opts *WorkloadOptions) PublishLocalSource(ctx context.Context, c *cli.Conf
...
-
 	currentRegistryOpts := source.RegistryOpts{CACertPaths: opts.CACertPaths, RegistryUsername: opts.RegistryUsername, RegistryPassword: opts.RegistryPassword, RegistryToken: opts.RegistryToken}
+	resitryWithProgress, err := source.GetRegistryWithProgress(ctx, &currentRegistryOpts, c.Stdout)
+	if err != nil {
+		return okToPush, err
+	}
 	ctx = logger.StashSourceImageLogger(ctx, logger.NewNoopLogger())
-
-	digestedImage, err := source.ImgpkgPush(ctx, contentDir, fileExclusions, &currentRegistryOpts, taggedImage)
+	digestedImage, err := source.ImgpkgPush(ctx, contentDir, fileExclusions, resitryWithProgress, taggedImage)
```
Comment (shashwathi) [inline pkg/commands/workload.go:438]: resitryWithProgress
```suggestion
	digestedImage, err := source.ImgpkgPush(ctx, contentDir, fileExclusions, registryWithProgress, taggedImage)
```

File: pkg/commands/workload.go [hunk 2]
```diff
@@ -429,11 +429,13 @@ func (opts *WorkloadOptions) PublishLocalSource(ctx context.Context, c *cli.Conf
 	}
 
 	c.Infof("Publishing source in %q to %q...\n", opts.LocalPath, taggedImage)
-
```
Comment (shashwathi) [inline pkg/commands/workload.go:432]: do not delete new line

File: pkg/commands/workload_test.go [hunk 3]
```diff
@@ -1805,6 +1805,8 @@ func TestWorkloadOptionsPublishLocalSourcePrivateRegistry(t *testing.T) {
 	u, err := url.Parse(reg.URL)
 	utilruntime.Must(err)
 	registryHost := u.Host
+	levelLogger := logger.NewSinkLogger("", source.Int32Ptr(1), nil)
+	testprogresslogger := logger.NewNoopProgressBar(levelLogger)
```
Comment (shashwathi) [inline pkg/commands/workload_test.go:1809]:
```suggestion
	testProgressLogger := logger.NewNoopProgressBar(levelLogger)
```

File: pkg/commands/workload_test.go [hunk 4]
```diff
@@ -1867,7 +1869,7 @@ Published source
 			opts.LoadDefaults(c)
 			opts.DefineFlags(ctx, c, cmd)
 			cmd.ParseFlags(test.args)
-
+			ctx = logger.StashProgressBarLogger(ctx, testprogresslogger)
```
Comment (shashwathi) [inline pkg/commands/workload_test.go:1872]: any reason why its not loaded into context right after `testProgressLogger` is declared

File: pkg/logger/progress_logger_test.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+	"github.com/google/go-cmp/cmp"
+	regv1 "github.com/google/go-containerregistry/pkg/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/vmware-tanzu/apps-cli-plugin/pkg/cli-runtime"
+	"github.com/vmware-tanzu/apps-cli-plugin/pkg/logger"
+)
+
+func TestProgressLogger(t *testing.T) {
+	t.Run("Test progress logger", func(t *testing.T) {
```
Comment (shashwathi) [inline pkg/logger/progress_logger_test.go:35]: Why is t.Run is used here?

File: pkg/logger/progress_logger_test.go [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+		outputwriter := &bytes.Buffer{}
+		scheme := runtime.NewScheme()
+		c := cli.NewDefaultConfig("test", scheme)
+		c.Stdout = outputwriter
+		levelLogger := logger.NewSinkLogger("", int32Ptr(1), nil)
+		testprogresslogger := logger.NewNoopProgressBar(levelLogger)
```
Comment (shashwathi) [inline pkg/logger/progress_logger_test.go:41]: camelcase for variable names

File: pkg/logger/progress_logger_test.go [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+		processData(testprogresslogger)
```
Comment (shashwathi) [inline pkg/logger/progress_logger_test.go:42]: why is the func created for this? 
Consider gofunc to create and pass info to channel.

File: pkg/logger/progress_logger.go [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+	return &ProgressBarLogger{logger: logger, finalMessage: finalMessage, errorMessagePrefix: errorMessagePrefix, writer: writer}
+}
+
+// NewNoopProgressBar constructs a Noop Progress bar that will not display anything
+func NewNoopProgressBar(logger logr.Logger) ProgressLogger {
+	return &ProgressBarNoTTYLogger{logger: logger}
+}
+
+// ProgressBarLogger display progress bar on output
+type ProgressBarLogger struct {
```
Comment (shashwathi) [inline pkg/logger/progress_logger.go:47]: No unit tests on the actual progress bar

Comment (shashwathi) [inline pkg/logger/progress_logger.go:47]: Rename the struct to `ProgressBar` to avoid some confusion

File: pkg/logger/progress_logger_test.go [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+	"github.com/vmware-tanzu/apps-cli-plugin/pkg/cli-runtime"
+	"github.com/vmware-tanzu/apps-cli-plugin/pkg/logger"
+)
+
+func TestProgressLogger(t *testing.T) {
+	t.Run("Test progress logger", func(t *testing.T) {
+		outputwriter := &bytes.Buffer{}
+		scheme := runtime.NewScheme()
+		c := cli.NewDefaultConfig("test", scheme)
```
Comment (shashwathi) [inline pkg/logger/progress_logger_test.go:38]: why is this required?

File: pkg/logger/progress_logger_test.go [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+		c.Stdout = outputwriter
```
Comment (shashwathi) [inline pkg/logger/progress_logger_test.go:39]: why is this code required for the test

Comment (shashwathi) [inline pkg/logger/progress_logger_test.go:39]: why is this code required for the test

File: pkg/logger/uilogger_test.go [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+	"github.com/spf13/cobra"
+
+	"github.com/vmware-tanzu/apps-cli-plugin/pkg/cli-runtime"
+	"github.com/vmware-tanzu/apps-cli-plugin/pkg/logger"
+)
+
+func TestNoopLogger(t *testing.T) {
+	t.Run("NoopLogger", func(t *testing.T) {
+		cmd := &cobra.Command{}
+		ctx := cli.WithCommand(context.Background(), cmd)
```
Comment (shashwathi) [inline pkg/logger/uilogger_test.go:32]: you do not need cmd in context

File: pkg/logger/uilogger_test.go [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+		if noopLogger == nil {
+			t.Errorf("Stash or Retrieve NoopLogger Failed")
+		}
+	})
+}
+
+func TestRetrieveProgressLogger(t *testing.T) {
+	t.Run("NoopLogger", func(t *testing.T) {
+		cmd := &cobra.Command{}
+		ctx := cli.WithCommand(context.Background(), cmd)
```
Comment (shashwathi) [inline pkg/logger/uilogger_test.go:44]: you do not need cmd in context

File: pkg/commands/workload_test.go (hunk 4 above)
Comment (shaheerkootteeri) [reply pkg/commands/workload_test.go:1872]: I wanted to reuse the same Progress bar logger instance for all test rather than redefining for each tests

File: pkg/logger/progress_logger_test.go (hunk 5 above)
Comment (shaheerkootteeri) [reply pkg/logger/progress_logger_test.go:35]: I had table test cases earlier , and had race conditions trying to test the actual progress bar. Removed the table test, but didn't address the test run function. I will update the code

File: pkg/logger/progress_logger.go [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+func (l *ProgressBarLogger) End() {
+	if l.cancelFunc != nil {
+		l.cancelFunc()
+	}
+	l.bar.Finish()
+	l.logger.Info(l.finalMessage)
+}
+
+// ProgressBarNoTTYLogger does not display the progress bar
+type ProgressBarNoTTYLogger struct {
```
Comment (shashwathi) [inline pkg/logger/progress_logger.go:101]: Make the noop progress bar logger simpler. Remove `finalMessage` and also logger. Its meant to sub in as something simple for higher level tests

File: pkg/logger/progress_logger.go (hunk 8 above)
Comment (shaheerkootteeri) [reply pkg/logger/progress_logger.go]: I couldn't add unit test cases as it is causing race conditions. If we have the out as `os.stdout / os.stderr `instead of `bytes.Buffer ` , the progress bar works fine. However reading back from os.stdout to valid is challenging.

File: pkg/logger/uilogger_test.go (hunk 11 above)
Comment (shaheerkootteeri) [reply pkg/logger/uilogger_test.go]: I have updated the test case to remove the command

File: pkg/logger/uilogger_test.go (hunk 12 above)
Comment (shaheerkootteeri) [reply pkg/logger/uilogger_test.go]: I have updated the test case to remove the command

File: pkg/logger/progress_logger.go [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+}
+
+// Start the display of the Progress Bar
+func (l *ProgressBar) Start(ctx context.Context, progressChan <-chan regv1.Update) {
+	ctx, cancelFunc := context.WithCancel(ctx)
+	l.cancelFunc = cancelFunc
+	// Add a new empty line to separate the progress bar from prior output
+	fmt.Println()
+	l.bar = pb.New64(0)
+	l.bar.SetWriter(l.writer)
```
Comment (shashwathi) [inline pkg/logger/progress_logger.go:59]: delete this line. you do not need this.

File: pkg/logger/progress_logger.go [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+	return &ProgressBar{writer: writer}
+}
+
+// NewNoopProgressBar constructs a Noop Progress bar that will not display anything
+func NewNoopProgressBar() ProgressLogger {
+	return &ProgressBarNoTTY{}
+}
+
+// ProgressBarLogger display progress bar on output
+type ProgressBar struct {
```
Comment (shashwathi) [inline pkg/logger/progress_logger.go:46]: make this struct private. its not accessed outside of the package

File: pkg/logger/progress_logger.go [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+
+// ProgressLogger interface should match with imgpkg util Progress logger
+type ProgressLogger interface {
+	Start(ctx context.Context, progress <-chan regv1.Update)
+	End()
+}
+
+// NewProgressBar constructor to build a ProgressLogger responsible for printing out a progress bar using updates when
+// writing to a registry via ggcr
+func NewProgressBar(writer io.Writer) ProgressLogger {
```
Comment (shashwathi) [inline pkg/logger/progress_logger.go:36]: create a folder `fake` under logger. Dump `NewNoopProgressBar` related code into that. This will clearly separate testing fakes and actual code.

File: pkg/logger/progress_logger.go [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+
+// ProgressBarLogger display progress bar on output
+type ProgressBar struct {
+	cancelFunc context.CancelFunc
+	bar        *pb.ProgressBar
+	writer     io.Writer
+}
+
+// Start the display of the Progress Bar
+func (l *ProgressBar) Start(ctx context.Context, progressChan <-chan regv1.Update) {
```
Comment (shashwathi) [inline pkg/logger/progress_logger.go:53]: add unit tests for `progressBar`  struct. remove tests for noop progress bar

File: pkg/source/registry_test.go [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+
+import (
+	"bytes"
+	"context"
+	"testing"
+
+	"github.com/vmware-tanzu/apps-cli-plugin/pkg/source"
+)
+
+func TestGetRegistryWithProgress(t *testing.T) {
```
Comment (shashwathi) [inline pkg/source/registry_test.go:27]: not sure what value this is adding, feel free to delete this test file

**Review by shashwathi — CHANGES_REQUESTED** [review]

File: pkg/source/registry_test.go (hunk 18 above)
Comment (shaheerkootteeri) [reply pkg/source/registry_test.go:27]: deleted

File: pkg/logger/progress_logger.go (hunk 17 above)
Comment (shaheerkootteeri) [reply pkg/logger/progress_logger.go]: Done

File: pkg/logger/progress_logger.go (hunk 16 above)
Comment (shaheerkootteeri) [reply pkg/logger/progress_logger.go:36]: Done

File: pkg/logger/progress_logger.go (hunk 15 above)
Comment (shaheerkootteeri) [reply pkg/logger/progress_logger.go]: Done for both Noop as well as ProgressBar

File: pkg/logger/progress_logger.go (hunk 14 above)
Comment (shaheerkootteeri) [reply pkg/logger/progress_logger.go]: Done

File: pkg/source/registry.go [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+type RegistryOpts struct {
+	CACertPaths      []string
+	RegistryUsername string
+	RegistryPassword string
+	RegistryToken    string
+}
+
+// GetRegistryWithProgress creates new registry instance that provides
+// progress updates to the logger
+func GetRegistryWithProgress(ctx context.Context, registryOpts *RegistryOpts) (*registry.WithProgress, error) {
```
Comment (odinnordico) [inline pkg/source/registry.go:38]: As you are not getting an existing resource change the name to `NewRegistryWithProgress` or `CreateRegistryWithProgress`

File: pkg/logger/progress_logger.go [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+// NewProgressBar constructor to build a ProgressLogger responsible for printing out a progress bar using updates when
+// writing to a registry via ggcr
+func NewProgressBar() ProgressLogger {
+	return &progressBar{}
+}
+
+// ProgressBarLogger display progress bar on output
+type progressBar struct {
```
Comment (odinnordico) [inline pkg/logger/progress_logger.go:40]: Comment and function name should match

File: pkg/logger/progress_logger.go [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+					l.bar.Start()
+				}
+				l.bar.SetCurrent(update.Complete)
+				l.bar.Write()
+			}
+		}
+	}()
+}
+
+// End stops the progress bar and writes the final message
```
Comment (odinnordico) [inline pkg/logger/progress_logger.go:77]: Does `l.cancelFunc()` or `l.bar.Finish()` update writes a final message? if not update the comment

File: testing/e2e/workload_test.go [hunk 22]
```diff
@@ -181,6 +182,10 @@ func TestCreateFromGitWithAnnotations(t *testing.T) {
 				dir, _ := ioutil.TempDir("", "")
 				defer os.RemoveAll(dir)
 
+				if regexProgress.FindString(output) == "" {
+					t.Errorf("unexpected error: Expected ProgressBar in output")
```
Comment (odinnordico) [inline testing/e2e/workload_test.go:186]: add the `output` in the error message

File: testing/e2e/workload_test.go [hunk 23]
```diff
@@ -330,6 +339,10 @@ func TestCreateFromGitWithAnnotations(t *testing.T) {
...
```
Comment (odinnordico) [inline testing/e2e/workload_test.go:343]: add the `output` in the error message

File: testing/e2e/workload_test.go [hunk 24]
```diff
@@ -45,9 +45,10 @@ var (
 )
 
 var (
-	pattern    = "\\\\U000[a-zA-Z0-9]+"
-	regexEmoji = regexp.MustCompile(pattern)
-	regexPod   = regexp.MustCompile("\\s*NAME\\s+READY\\s+STATUS\\s+RESTARTS\\s+AGE\\s*")
+	pattern       = "\\\\U000[a-zA-Z0-9]+"
+	regexEmoji    = regexp.MustCompile(pattern)
+	regexPod      = regexp.MustCompile("\\s*NAME\\s+READY\\s+STATUS\\s+RESTARTS\\s+AGE\\s*")
+	regexProgress = regexp.MustCompile(`[0-9]*\.*[0-9]*\s[a-zA-Z]+\s/\s[0-9]*[\.]*[0-9]*\s[a-zA-Z]+\s\[-+>*\]\s[0-9]*\.[0-9]+%.*`)
```
Comment (odinnordico) [inline testing/e2e/workload_test.go:51]: lets refine the regex a bit to match better the progress output

File: pkg/logger/progress_logger_test.go [hunk 25]
```diff
@@ -0,0 +1,96 @@
...
+	"os"
+	"regexp"
+	"testing"
+	"time"
+
+	regv1 "github.com/google/go-containerregistry/pkg/v1"
+)
+
+func TestProgressBar(t *testing.T) {
+	var regexProgress = regexp.MustCompile(`[0-9]*\.*[0-9]*\s[a-zA-Z]+\s/\s[0-9]*[\.]*[0-9]*\s[a-zA-Z]+\s\[-+>*\]\s[0-9]*\.[0-9]+%.*`)
```
Comment (odinnordico) [inline pkg/logger/progress_logger_test.go:33]: lets refine the regex a bit to match better the progress output

Comment (shaheerkootteeri) [reply pkg/logger/progress_logger_test.go:33]: The regex is matching the progress output

File: testing/e2e/workload_test.go (hunk 24 above)
Comment (shaheerkootteeri) [reply testing/e2e/workload_test.go:51]: The regex is matching the progress output

File: pkg/logger/fake/fake_progress_logger.go [hunk 26]
```diff
@@ -0,0 +1,41 @@
+package fake
+
```
Comment (warango4) [inline pkg/logger/fake/fake_progress_logger.go:2]: Should this file have Copyright?

File: pkg/logger/progress_logger_test.go [hunk 27]
```diff
@@ -0,0 +1,96 @@
...
+	tests := []struct {
+		name      string
+		regObject []regv1.Update
+	}{{
+		name:      "ProgressLogger",
+		regObject: []regv1.Update{{Total: 100, Complete: 100, Error: nil}},
+	}, {
+		name:      "ProgressLogger with Error",
+		regObject: []regv1.Update{{Total: 100, Complete: 20, Error: fmt.Errorf("test Error")}, {Total: 100, Complete: 100, Error: nil}},
```
Comment (shashwathi) [inline pkg/logger/progress_logger_test.go:42]: add test cases for 
- when total is set to 0 , it should print no progress bar
- when context is closed before End is called, progress bar should stop mid way and exit.

File: pkg/logger/fake/fake_progress_logger.go [hunk 28]
```diff
@@ -0,0 +1,57 @@
...
+)
+
+// NewNoopProgressBar constructs a Noop Progress bar that will not display anything
+func NewNoopProgressBar() logger.ProgressLogger {
+	return &progressBarNoTTY{}
+}
+
+// ProgressBarNoTTYLogger does not display the progress bar
+type progressBarNoTTY struct {
+	cancelFunc context.CancelFunc
```
[the remark this answers is not in the document]
Comment (shaheerkootteeri) [reply pkg/logger/fake/fake_progress_logger.go:34]: This gets called when local path tests run as part of workload_test.go file. Removing it might not abandon all the loggers.

File: pkg/logger/fake/fake_progress_logger.go [hunk 29]
```diff
@@ -0,0 +1,57 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fake
+
+import (
+	"context"
+
```
[the remark this answers is not in the document]
Comment (shaheerkootteeri) [reply pkg/logger/fake/fake_progress_logger.go:21]: the new line is being added by VS Code lint extensions. I can remove them if needed , however will be something that keeps coming back as a change every-time a new file changes and all the packages are rebuilt.

File: pkg/logger/progress_logger.go [hunk 30]
```diff
@@ -0,0 +1,97 @@
...
+	cancelFunc context.CancelFunc
+	bar        *pb.ProgressBar
+}
+
+// Start the display of the Progress Bar
+func (l *progressBar) Start(ctx context.Context, progressChan <-chan regv1.Update) {
+	ctx, cancelFunc := context.WithCancel(ctx)
+	l.cancelFunc = cancelFunc
+	// Add a new empty line to separate the progress bar from prior output
+	fmt.Println()
```
[the remark this answers is not in the document]
Comment (shaheerkootteeri) [reply pkg/logger/progress_logger.go:50]: If the ProgressBar hasn't been started for whatever reason , we wouldn't need the new line as part of the existing output, and would be a change in format for existing output. Hence I moved inside the Start function as this new line should happen only if there is a Start for progress logger. We could remove that new line entirely if we are worried about the real-estate.

File: testing/e2e/workload_test.go [hunk 31]
```diff
@@ -181,6 +182,10 @@ func TestCreateFromGitWithAnnotations(t *testing.T) {
 				dir, _ := ioutil.TempDir("", "")
 				defer os.RemoveAll(dir)
 
+				if regexProgress.FindString(output) == "" {
+					t.Errorf("unexpected error: expected progressbar in output %v", output)
```
Comment (shashwathi) [inline testing/e2e/workload_test.go:186]: repetition of words:
how about `t.Errorf("error: expected progressbar in output %v", output)`
`t.Errorf("expected progressbar in output %v", output)`

Comment (shashwathi) [reply testing/e2e/workload_test.go:186]: same for rest of the error msgs in this file

File: pkg/logger/progress_logger.go (hunk 20 above)
Comment (shashwathi) [reply pkg/logger/progress_logger.go:40]: +1

File: pkg/logger/fake/fake_progress_logger.go [hunk 32]
```diff
@@ -0,0 +1,57 @@
...
+	"github.com/vmware-tanzu/apps-cli-plugin/pkg/logger"
+)
+
+// NewNoopProgressBar constructs a Noop Progress bar that will not display anything
+func NewNoopProgressBar() logger.ProgressLogger {
+	return &progressBarNoTTY{}
+}
+
+// ProgressBarNoTTYLogger does not display the progress bar
+type progressBarNoTTY struct {
```
Comment (shashwathi) [inline pkg/logger/fake/fake_progress_logger.go:33]: func name and comment should match

File: testing/e2e/workload_test.go (hunk 31 above)
Comment (shaheerkootteeri) [reply testing/e2e/workload_test.go]: Updated

File: pkg/logger/progress_logger.go (hunk 30 above)
[the remark this answers is not in the document]
Comment (shaheerkootteeri) [reply pkg/logger/progress_logger.go]: Removed the new line based on the conversation with Daniel

File: pkg/logger/fake/fake_progress_logger.go (hunk 32 above)
Comment (shaheerkootteeri) [reply pkg/logger/fake/fake_progress_logger.go:33]: Updated

**Review by shashwathi — APPROVED** [review]

**Review by odinnordico — APPROVED** [review]
LGTM

**Review by warango4 — APPROVED** [review]
LGTM
